README
-------------------------------------------------------------------------------
Integrated control station for arbitrary APM-based unmanned vehicles
UxV-CS

(C) 2012 by Stefan Gofferje
http://stefan.gofferje.net/uxv-cs

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
-------------------------------------------------------------------------------

This code is currently under heavy development and is far from productive
use yet.

CHANGELOG
Integrated control station for arbitrary APM-based unmanned vehicles
UxV-CS

(C) 2012 by Stefan Gofferje
http://stefan.gofferje.net/uxv-cs

Changelog
--------------------------------------------------------------------

---- 24feb2013
- started work on systems settings
- added number of LiPo cells setting for alarm thresholds (EEPROM)
- added LiPo voltage per cell warning threshold (EEPROM)
- added LiPo voltage per cell alarm threshold (EEPROM)

---- 23feb2013
- moved FrSky telemetry decoder to own library

---- 21feb2013
- added FrSky telemetry timeout counter

---- 20feb2013
- started FrSky telemetry integration
- integrated RSSI indicators into status-bar

---- 17feb2013
- minor improvements and bugfixes

---- 10feb2013
- introduced saving of settings in EEPROM
- minor improvements and bugfixes
- worked on code readability
- added comments to code

---- 10feb2013
- added metric, imperial land and nautic display modes
- changed PFD VSI from numerical to graphical
- added status bar indicator for HEARTBEAT.custom_mode
- minor fixes
- removed all uses of String object from sketch and eDIPTFT lib
- new methods for eDIPTFT lib

---- 05feb2013
- PFD ready
- Overview page almost ready
- New functions for the EDIPTFT lib
- bugfixes
- publishes raw graphics for instruments under CC-BY-SA 3.0

---- 03feb2013
- mavlink_msg_request_data_stream implemented and WORKS
- LCD comms switched to smallprotocol

---- 01feb2013
- Basic MAVLink functionality implemented
- Basic display framework implemented

---- 13jan2013
Project start


arduino/libraries/EDIPTFT/EDIPTFT.h
//
// Library for controlling Electronic Assembly eDIPTFT displays
//
//      Copyright (c) 2013 Stefan Gofferje. All rights reserved.
//
//      This library is free software; you can redistribute it and/or
//      modify it under the terms of the GNU Lesser General Public
//      License as published by the Free Software Foundation; either
//      version 2.1 of the License, or (at your option) any later
//      version.
//
//      This library is distributed in the hope that it will be
//      useful, but WITHOUT ANY WARRANTY; without even the implied
//      warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//      PURPOSE.  See the GNU Lesser General Public License for more
//      details.
//
//      You should have received a copy of the GNU Lesser General
//      Public License along with this library; if not, write to the
//      Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
//      Boston, MA 02110-1301 USA
//

#ifndef EDIPTFT_h
#define EDIPTFT_h

#define EA_BLACK 1
#define EA_BLUE 2
#define EA_RED 3
#define EA_GREEN 4
#define EA_PURPLE 5
#define EA_CYAN 6
#define EA_YELLOW 7
#define EA_WHITE 8
#define EA_DARKGREY 9
#define EA_ORANGE 10
#define EA_LILA 11
#define EA_DARKPURPLE 12
#define EA_MINT 13
#define EA_GRASSGREEN 14
#define EA_LIGHTBLUE 15
#define EA_LIGHTGREY 16

#define NAK 0x15
#define ACK 0x06

class EDIPTFT {
  public:
    EDIPTFT(int port, int smallprotocol);
    char readByte();
    char waitandreadByte();
    unsigned char datainBuffer();
    void readBuffer(char* data);
    void smallProtoSelect(char address);
    void smallProtoDeselect(char address);
    void sendData(char* data, char len);
    void clear();
    void invert();
    void setDisplayColor(char fg, char bg);
    void fillDisplayColor(char bg);
    void terminalOn(boolean on);
    void cursor(boolean on);
    void setCursor(char col, char row);
    void defineBargraph(char dir, char no, int x1, int y1, int x2, int y2, byte sv, byte ev, char type);
    void updateBargraph(char no, char val);
    void setBargraphColor(char no, char fg, char bg, char fr);
    void makeBargraphTouch(char no);
    void linkBargraphLight(char no);
    void deleteBargraph(char no,char n1);
    void defineInstrument(char no, int x1, int y1, char image, char angle, char sv, char ev);
    void updateInstrument(char no, char val);
    void redrawInstrument(char no);
    void deleteInstrument(char no, char n1, char n2);
    void setTextColor(char fg, char bg);
    void setTextFont(char font);
    void setTextAngle(char angle);
    void drawText(int x1, int y1, char justification,char* text);
    void setLineColor(char fg, char bg);
    void setLineThick(char x, char y);
    void drawLine(int x1, int y1, int x2, int y2);
    void drawRect(int x1, int y1, int x2, int y2);
    void drawRectf(int x1, int y1, int x2, int y2, char color);
    void defineTouchKey(int x1, int y1, int x2, int y2, char down, char up, char* text);
    void setTouchkeyColors(char n1, char n2, char n3, char s1, char s2, char s3);
    void setTouchkeyFont(char font);
    void setTouchkeyLabelColors(char nf,char sf);
    void removeTouchArea(char code,char n1);
  private:
    int _port;
    int _smallprotocol;
    unsigned char bytesAvailable();
    void waitBytesAvailable();
    void sendByte(char data);
    void sendSmall(char* data, char len);
    void sendSmallDC2(char* data, char len);
};
#endif

arduino/libraries/EDIPTFT/EDIPTFT.cpp
//
// Library for controlling Electronic Assembly eDIPTFT displays
//
//      Copyright (c) 2013 Stefan Gofferje. All rights reserved.
//
//      This library is free software; you can redistribute it and/or
//      modify it under the terms of the GNU Lesser General Public
//      License as published by the Free Software Foundation; either
//      version 2.1 of the License, or (at your option) any later
//      version.
//
//      This library is distributed in the hope that it will be
//      useful, but WITHOUT ANY WARRANTY; without even the implied
//      warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//      PURPOSE.  See the GNU Lesser General Public License for more
//      details.
//
//      You should have received a copy of the GNU Lesser General
//      Public License along with this library; if not, write to the
//      Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
//      Boston, MA 02110-1301 USA
//

#include <FastSerial.h>
#include <EDIPTFT.h>

#define DEBUG false

EDIPTFT::EDIPTFT(int port, int smallprotocol) {
  _port = port;
  _smallprotocol = smallprotocol;
}

void EDIPTFT::sendByte(char data) {
  switch (_port) {
    case 0 : {
      Serial.write(data);
      break;
    }
    case 1 : {
      Serial1.write(data);
      break;
    }
    case 2 : {
      Serial2.write(data);
      break;
    }
    case 3 : {
      Serial3.write(data);
      break;
    }
  }
}  

char EDIPTFT::readByte() {
  char result;
  switch (_port) {
    case 0 : {
      result = Serial.read();
      break;
    }
    case 1 : {
      result = Serial1.read();
      break;
    }
    case 2 : {
      result = Serial2.read();
      break;
    }
    case 3 : {
      result = Serial3.read();
      break;
    }
  }
  return result;
}  

void EDIPTFT::waitBytesAvailable() {
  while (bytesAvailable() == 0) {
  }
}

char EDIPTFT::waitandreadByte() {
  waitBytesAvailable();
  char result = readByte();
  return(result);
}
  

unsigned char EDIPTFT::bytesAvailable() {
  unsigned char result;
  switch (_port) {
    case 0 : {
      result = Serial.available();
      break;
    }
    case 1 : {
      result = Serial1.available();
      break;
    }
    case 2 : {
      result = Serial2.available();
      break;
    }
    case 3 : {
      result = Serial3.available();
      break;
    }
  }
  return result;
}  

void EDIPTFT::sendData(char* data, char len) {
  if (DEBUG) {
    char i;
    for (i=0;i<len;i++) {
      Serial.print(byte(data[i]),HEX);
      Serial.print(" ");
    }
    Serial.println();
  }  
    
  if (_smallprotocol > 0) {
    sendSmall(data,len);
  }
  else {
    unsigned char i;
    for(i=0; i < len; i++) {
      sendByte(data[i]);
    }
  }
}

void EDIPTFT::sendSmall(char* data, char len) {
  unsigned char i, bcc;
  char ok = 0;
  
  while (ok == 0) {
    sendByte(0x11);
    bcc = 0x11;
    
    sendByte(len);
    bcc = bcc + len;
    
    for(i=0; i < len; i++) {
      sendByte(data[i]);
      bcc = bcc + data[i];
    }
    sendByte(bcc);
    waitBytesAvailable();
    if (bytesAvailable() > 0) {
      if (readByte() == ACK) ok = 1;
      else {
	ok = 0;
      }
    }
    else {
      delay(2000);
      ok = 0;
    }
  }
}

void EDIPTFT::sendSmallDC2(char* data, char len) {
  unsigned char i, bcc;
  char ok = 0;
  
  while (ok == 0) {
  
    sendByte(0x12);
    bcc = 0x12;
    
    for(i=0; i < len; i++) {
      sendByte(data[i]);
      bcc = bcc + data[i];
    }
    sendByte(bcc);
    delay(6);
    if (bytesAvailable() > 0) {
      if (readByte() == ACK) ok = 1;
      else ok = 0;
    }
    else {
      delay(2000);
      ok = 0;
    }
  }
}

void EDIPTFT::smallProtoSelect(char address) {
  char command [] = {
    0x03,'A','S',address
  };
  sendSmallDC2(command,4);
}  

void EDIPTFT::smallProtoDeselect(char address) {
  char command [] = {
    0x03,'A','D',address
  };
  sendSmallDC2(command,4);
}  

unsigned char EDIPTFT::datainBuffer() {
  unsigned char result;
  char command [] = {
    0x01,'I'
  };
  sendSmallDC2(command,2);
  waitandreadByte();
  waitandreadByte();
  result=waitandreadByte();
  waitandreadByte();
  waitandreadByte();
  return result;
}  

void EDIPTFT::readBuffer(char* data) {
  char len,i;
  char command [] = {
    0x01,'S'
  };
  sendSmallDC2(command,2);
  waitandreadByte();
  len=waitandreadByte();
  char result[len];
  for (i=0;i<len;i++) {
    result[i] = waitandreadByte();
  }
  memcpy(data,result,len);
  waitandreadByte();
}
  

void EDIPTFT::clear() {
  char command [] = {12};
  sendData(command,1);
}

void EDIPTFT::invert() {
  char command [] = {
    27,'D','I'
  };
  sendData(command,3);
}

void EDIPTFT::setDisplayColor(char fg, char bg) {
  char command [] = {
    27,'F','D',fg,bg
  };
  sendData(command,5);
}

void EDIPTFT::fillDisplayColor(char bg) {
  char command [] = {
    27,'D','F',bg
  };
  sendData(command,4);
}

void EDIPTFT::terminalOn(boolean on) {
  if (on) {
    char command [] = {27,'T','E'};
    sendData(command,3);
  }
  else {
    char command [] = {27,'T','A'};
    sendData(command,3);
  }
}

void EDIPTFT::cursor(boolean on) {
  if (on) {
    char command [] = {27,'T','C',1};
    sendData(command,4);
  }
  else {
    char command [] = {27,'T','C',0};
    sendData(command,4);
  }
}

void EDIPTFT::setCursor(char col, char row) {
  char command [] = {27,'T','P',col,row};
  sendData(command,5);
}

void EDIPTFT::defineBargraph(char dir, char no, int x1, int y1, int x2, int y2, byte sv, byte ev, char type) {
  char command [] = {
    27,'B',dir,no,
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
    lowByte(x2),highByte(x2),lowByte(y2),highByte(y2),
    char(sv),
    char(ev),
    type
  };
  sendData(command,15);
}

void EDIPTFT::updateBargraph(char no, char val) {
  char command [] = {
    27,'B','A',no,val
  };
  sendData(command,5);
}

void EDIPTFT::setBargraphColor(char no, char fg, char bg, char fr) {
  char command [] = {
    27,'F','B',no,fg,bg,fr
  };
  sendData(command,7);
}

void EDIPTFT::linkBargraphLight(char no) {
  char command [] = {
    27,'Y','B',no
  };
  sendData(command,4);
}  

void EDIPTFT::makeBargraphTouch(char no) {
  char command [] = {
    27,'A','B',no
  };
  sendData(command,4);
}  

void EDIPTFT::deleteBargraph(char no,char n1) {
  char command [] = {
    27,'B','D',no,n1
  };
  sendData(command,5);
}

void EDIPTFT::defineInstrument(char no, int x1, int y1, char image, char angle, char sv, char ev) {
  char command [] = {
    27,'I','P',no,
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
    image,angle,sv,ev
  };
  sendData(command,12);
}

void EDIPTFT::updateInstrument(char no, char val) {
  char command [] = {
    27,'I','A',no,val
  };
  sendData(command,5);
}

void EDIPTFT::redrawInstrument(char no) {
  char command [] = {
    27,'I','N',no
  };
  sendData(command,4);
}

void EDIPTFT::deleteInstrument(char no, char n1, char n2) {
  char command [] = {
    27,'B','D',no,n1,n2
  };
  sendData(command,6);
}

void EDIPTFT::setLineColor(char fg, char bg) {
  char command [] = {
    27,'F','G',fg,bg
  };
  sendData(command,5);
}

void EDIPTFT::setLineThick(char x, char y) {
  char command [] = {
    27,'G','Z',x,y
  };
  sendData(command,5);
}

void EDIPTFT::setTextColor(char fg, char bg) {
  char command [] = {
    27,'F','Z',fg,bg
  };
  sendData(command,5);
}

void EDIPTFT::setTextFont(char font) {
  char command [] = {
    27,'Z','F',font
  };
  sendData(command,4);
}

void EDIPTFT::setTextAngle(char angle) {
  // 0 = 0°, 1 = 90°, 2 = 180°, 3 = 270°
  char command [] = {
    27,'Z','W',angle
  };
  sendData(command,4);
}

void EDIPTFT::drawText(int x1, int y1, char justification,char* text) {
  byte len = strlen(text);
  byte i;
  char helper [len+8];
  char command [] = {
    27,'Z',justification,
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
  };
  for (i=0;i<=6;i++) helper[i] = command[i];
  for (i=0;i<=len;i++) helper[i+7] = text[i];
  sendData(helper,len+8);
}

void EDIPTFT::drawLine(int x1, int y1, int x2, int y2) {
  char command [] = {
    27,'G','D',
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
    lowByte(x2),highByte(x2),lowByte(y2),highByte(y2),
  };
  sendData(command,11);
}

void EDIPTFT::drawRect(int x1, int y1, int x2, int y2) {
  char command [] = {
    27,'G','R',
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
    lowByte(x2),highByte(x2),lowByte(y2),highByte(y2),
  };
  sendData(command,11);
}

void EDIPTFT::drawRectf(int x1, int y1, int x2, int y2, char color) {
  char command [] = {
    27,'R','F',
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
    lowByte(x2),highByte(x2),lowByte(y2),highByte(y2),
    color
  };
  sendData(command,12);
}

void EDIPTFT::defineTouchKey(int x1, int y1, int x2, int y2, char down, char up, char* text) {
  byte len = strlen(text);
  byte i;
  char helper [len+13];
  char command [] = {
    27,'A','T',
    lowByte(x1),highByte(x1),lowByte(y1),highByte(y1),
    lowByte(x2),highByte(x2),lowByte(y2),highByte(y2),
    down,up
  };
  for (i=0;i<=12;i++) helper[i] = command[i];
  for (i=0;i<=len;i++) helper[i+13] = text[i];

  sendData(helper,len+14);
}

void EDIPTFT::setTouchkeyColors(char n1, char n2, char n3, char s1, char s2, char s3) {
  char command [] = {
    27,'F','E',n1,n2,n3,s1,s2,s3
  };
  sendData(command,9);
}

void EDIPTFT::setTouchkeyFont(char font) {
  char command [] = {
    27,'A','F',font
  };
  sendData(command,4);
}

void EDIPTFT::setTouchkeyLabelColors(char nf,char sf) {
  char command [] = {
    27,'F','A',nf,sf
  };
  sendData(command,5);
}

void EDIPTFT::removeTouchArea(char code,char n1) {
  char command [] = {
    27,'A','L',code,n1
  };
  sendData(command,5);
}




